package/resource_monitor/resource_monitor/src/query_process_file_descriptor.c

```diff
@@ -0,0 +1,491 @@
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#define _GNU_SOURCE
+
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+#include <search.h>
+
+#include <libsbp/linux.h>
+#include <dirent.h>
+#include <libpiksi/logging.h>
+#include <libpiksi/util.h>
+#include <libpiksi/runit.h>
+#include <libpiksi/table.h>
+
+#include "query_sys_state.h"
+#include "resource_monitor.h"
+#include "resource_query.h"
+#include "resmon_common.h"
+
+//#define DEBUG_QUERY_FD_TAB
+
+#define ITEM_COUNT 10
+#define MAX_PROCESS_COUNT 1024
+#define MAX_FILE_COUNT 4096
+#define SBP_FRAMING_MAX_PAYLOAD_SIZE 255
+#define LSOF_CMD_BUF_LENGTH 1024 * 128
+typedef enum {
+  SEND_FD_COUNTS_0,
+  SEND_FD_COUNTS_1,
+  SEND_FD_COUNTS_2,
+  SEND_FD_COUNTS_3,
+  SEND_FD_COUNTS_4,
+  SEND_FD_COUNTS_5,
+  SEND_FD_COUNTS_6,
+  SEND_FD_COUNTS_7,
+  SEND_FD_COUNTS_8,
+  SEND_FD_COUNTS_9,
+  SEND_FD_SUMMARY,
+  SEND_DONE,
+} send_state_t;
+
+typedef struct {
+  char id_str[16];
+  u16 pid;
+  u16 fd_count;
+} pid_entry_t;
+
+typedef struct {
+  char file_str[128];
+  u16 fd_count;
+} file_entry_t;
+
+MAKE_TABLE_WRAPPER(pid, pid_entry_t)
+MAKE_TABLE_WRAPPER(file, file_entry_t)
+
+typedef struct {
+  send_state_t send_state;
+  u32 total_fd_count;
+  table_t *pid_table;
+  table_t *file_table;
+  const char *top10_pid_counts[ITEM_COUNT];
+  const char *top10_file_counts[ITEM_COUNT];
+} resq_state_t;
+
+typedef struct id_to_count {
+  u16 count;
+  const char *id;
+} id_to_count_t;
+
+enum {
+  FD_STATE_ID_STR = 0,
+  FD_STATE_EXE_FILE_STR,
+  FD_STATE_FD_STR,
+  FD_STATE_DONE,
+  FD_STATE_COUNT,
+};
+
+static bool is_real_file_path(const char *path)
+{
+  if (strncmp(path, "pipe", 4) == 0 || strncmp(path, "socket", 6) == 0)
+    return false;
+  else
+    return true;
+}
+
+static bool process_fd_entry(resq_state_t *state, const char *file_str, const char *id_str)
+{
+  pid_entry_t *entry = pid_table_get(state->pid_table, id_str);
+  state->total_fd_count++;
+  if (entry == NULL) {
+    pid_entry_t new_entry = {0};
+    int s = snprintf(new_entry.id_str, sizeof(entry->id_str), "%s", id_str);
+
+    if (s < 0 || (size_t)s >= sizeof(entry->id_str)) {
+#ifdef DEBUG_QUERY_FD_TAB
+      PK_LOG_ANNO(LOG_ERR, "file descriptor query: failed to copy pid string");
+#endif
+      return false;
+    }
+
+    unsigned long pid = 0;
+    if (!strtoul_all(10, id_str, &pid)) {
+#ifdef DEBUG_QUERY_FD_TAB
+      PK_LOG_ANNO(LOG_ERR,
+                  "file descriptor query: failed to convert pid str to number: %s",
+                  strerror(errno));
+#endif
+      return false;
+    }
+
+    new_entry.pid = (u16)pid;
+
+    entry = calloc(1, sizeof(pid_entry_t));
+    *entry = new_entry;
+
+    bool added = pid_table_put(state->pid_table, id_str, entry);
+
+    if (!added) {
```

Comment:
I don't think we should disable this LOG_ERR call.  It means the table is full (so there's more than 1024 processes on the system) and we can no longer track all processes (meaning we can't potentially figure out if a process has a file handle leak associated with it).  So, suggestion is to not disable this and make the priority `LOG_ERR | LOG_SBP`.